### PR TITLE
Make all validate functions return MarineValidationResult

### PR DESCRIPTION
**Pull request by yehonatanz**

### CODE REVIEW

File: marine/marine.py [hunk 1]
```diff
@@ -166,22 +174,42 @@ def filter_and_parse(
...
+            return MarineValidationResult(valid, [err_msg_value.decode("utf-8")])
+        return MarineValidationResult(valid, [])
 
-    def validate_display_filter(self, display_filter: str) -> bool:
+    def validate_display_filter(self, display_filter: str) -> MarineValidationResult:
         display_filter = display_filter.encode("utf-8")
-        return bool(self._marine.validate_display_filter(display_filter))
+        err_msg = pointer(POINTER(c_char)())
+        valid = bool(self._marine.validate_display_filter(display_filter, err_msg))
+        if err_msg.contents:
```
Comment (shaharyarn) [inline marine/marine.py]: Maybe wrap those few lines by a functions that returns a tuple?

File: marine/marine.py [hunk 2]
```diff
@@ -228,7 +256,7 @@ def _add_or_get_filter(
         if filter_id < 0:
             err = None if err_msg_value is None else err_msg_value.decode("utf-8")
             if filter_id == c_int.in_dll(self._marine, "BAD_BPF_ERROR_CODE").value:
-                raise BadBPFException(err)
+                raise BadBPFException(f"Failed compiling the BPF: {err}")
```
Comment (shaharyarn) [inline marine/marine.py]: Why did you add extra text around this message? If its for clarity, why not to the other errors as well.

**Review by shaharyarn — CHANGES_REQUESTED** [review]
Add tests for the error of other validate functions as well (just to see it works, not for the syntax).

File: marine/marine.py (hunk 2 above)
Comment (yehonatanz) [reply marine/marine.py]: This prefix used to be added by our c core, I moved it here to make the c errors more uniform.
I'll remove it

File: tests/marine/test_marine.py [hunk 3]
```diff
@@ -961,9 +961,9 @@ def test_validate_fields_success(marine_instance: Marine):
 
 
 def test_validate_fields_failure(marine_instance: Marine):
-    assert not marine_instance.validate_fields(
-        ["ip.src", "eth.dst", "this.field.is.bad"]
-    )
+    result = marine_instance.validate_fields(["ip.src", "eth.dst", "this.field.is.bad"])
```
Comment (tomlegkov) [inline tests/marine/test_marine.py]: I think this test can be improved to better demonstrate the error behavior. For example, the array should contain a mix of good and bad fields `[good1, bad1, good2, bad2, ...]`. Does the error return all of the bad fields or does it fail after the first one? Is the order of the bad fields kept in the error message?

File: marine/marine.py [hunk 4]
```diff
@@ -164,24 +172,40 @@ def filter_and_parse(
         self._marine.marine_free(marine_result)
         return success, result
 
+    def _resolve_err_msg(self, err_msg: POINTER(POINTER(c_char))) -> Optional[str]:
+        if not err_msg.contents:
+            return None
+        error = string_at(err_msg.contents).decode("utf-8")
```
Comment (tomlegkov) [inline marine/marine.py]: Can `string_at` or `decode` fail for some reason? If so, should we put it in `try finally` to make sure the contents are always freed? I'd even consider adding a test to make sure memory is never leaked by the error message handler

**Review by tomlegkov — CHANGES_REQUESTED** [review]

File: marine/marine.py (hunk 4 above)
Comment (yehonatanz) [reply marine/marine.py]: `string_at` should not fail, since is marine core is guaranteed to provide a valid string there.
I guess it would be good to move the `decode` after the free though.

File: marine/marine.py (hunk 1 above)
Comment (yehonatanz) [reply marine/marine.py]: I decided to encapsulate only the handling of `err_msg`

**Review by tomlegkov — APPROVED** [review]

File: marine/marine.py [hunk 5]
```diff
@@ -18,6 +18,14 @@ class MarineResult(Structure):
 MARINE_NAME = "libmarine.so"
 
 
+class MarineValidationResult(NamedTuple):
```
Comment (tomlegkov) [inline marine/marine.py]: I feel this class should actually be 2 classes: `MarineFilterValidationResult` and `MarineFieldsValidationResult`.

I think that returning a list of errors when validating filters is a bit overkill since the list (by definition) will only have 1 item. What do you think?

File: tests/marine/test_marine.py [hunk 6]
```diff
@@ -928,7 +928,9 @@ def test_validate_bpf_success(marine_instance: Union[Marine, MarinePool]):
 
 
 def test_validate_bpf_failure(marine_instance: Union[Marine, MarinePool]):
-    assert not marine_instance.validate_bpf("what is this bpf?")
+    res = marine_instance.validate_bpf("what is this bpf?")
+    assert not res
+    assert "syntax error" in res.errors[0]
```
Comment (tomlegkov) [inline tests/marine/test_marine.py]: ^ This is what I meant by overkill. I think this API isn't the most intuitive when we only expect 1 error message that is a string

**Review by shaharyarn — APPROVED** [review]

File: marine/marine.py (hunk 5 above)
Comment (yehonatanz) [reply marine/marine.py]: As far as I recall, the old version of marine returns a `bool` and a `list` and I wanted to remain compatible.
Will make sure tomorrow.

Comment (yehonatanz) [reply marine/marine.py]: Checked.
Current version returns a `bool` and no error, so I took your advice and separated the classes :)

**Review by tomlegkov — APPROVED** [review]
